### PR TITLE
Add support for overriding resource requests

### PR DESCRIPTION
**Pull request by anjaberens**
Resource requests are currently not explicitly set - this results in k8s setting
the memory/cpu/disk requests to the same value as the limit.

For the most part, this is desired behaviour at Yelp as the main consumer of this
library is Tron and we don't expect our batches to need to "burst" any of these
resources. Additionally, we generally want to signal to k8s that our batches shouldn't
be killed if there are other (more interrupt-tolerant) workloads on an overloaded
node by defaulting the QoS class for any task_proc-created Pods to `Guaranteed`
precisely by setting `requests = limits`.

However, one of our other large task_proc workloads (Jolt) would like to have 
control over the requests as their workload runs in an isolated pool and cannot (at
least at the moment) be easily right-sized for optimal bin-packing/resource usage.

But, given that Jolt runs in its own pool, we can be a little cheeky and set lower
requests to pack Pods a little tighter since Jolt *is* quite interrupt-tolerant and
as such is resilient to its Pods getting killed due to node resource pressure (which
should be pretty rare in any case given the current usage patterns).

### CODE REVIEW

File: task_processing/plugins/kubernetes/kubernetes_pod_executor.py [hunk 1]
```diff
@@ -442,19 +444,38 @@ def _create_container_definition(
...
             )
 
+        # Set any optional requests
+        requests: Dict[str, Union[str, float]] = {}
+        if 'request_cpus' in task_config and task_config.request_cpus:
+            requests['cpus'] = task_config.request_cpus
+        if 'request_memory' in task_config and task_config.request_memory:
+            requests['memory'] = f"{task_config.request_memory}Mi"
+        if 'request_disk' in task_config and task_config.request_disk:
+            requests['ephemeral-storage'] = f"{task_config.request_disk}Mi"
```
Comment (nemacysts) [inline task_processing/plugins/kubernetes/kubernetes_pod_executor.py:454]: imo, it'd be nice to keep the disk stuff as-is since I don't think we have much of a system reservation for disk (if we have any at all) - but if this is just for completeness, it should be fine to leave :)

Comment (nemacysts) [inline task_processing/plugins/kubernetes/kubernetes_pod_executor.py:454]: if we set a default value of `None` for these, we should be able to avoid the `if FOO in task_config` bits here 


it's also maybe worth testing to see what happens if you feed k8s a null request as maybe we can build the `request` dict in the same way as the `limits` one - e.g.,
```python
requests = {
            "cpu": task_config.request_cpus if task_config.request_cpu else None,
            "memory": f"{task_config.request_memory}Mi"  if task_config.request_memory else None,
            "ephemeral-storage": f"{task_config.request_disk}Mi" if task_config.request_disk else None,
}
limits = {
            "cpu": task_config.cpus,
            "memory": f"{task_config.memory}Mi",
            "ephemeral-storage": f"{task_config.disk}Mi",
}
resources = V1ResourceRequirements(
                limits=limits,
                requests=requests,
)
```

File: task_processing/plugins/kubernetes/task_config.py [hunk 2]
```diff
@@ -304,6 +304,17 @@ def __invariant__(self) -> Tuple[Tuple[bool, str], ...]:
...
+
+    request_cpus = field(
+        type=float,
+        factory=float)
+    request_memory = field(
+        type=float,
+        factory=float)
+    request_disk = field(
+        type=float,
+        factory=float)
```
Comment (nemacysts) [inline task_processing/plugins/kubernetes/task_config.py:316]: imo, `$RESOURCE_request` would read a tad better :)

that said, for my suggestion above about giving these a null initial value:
```suggestion
    request_cpus = field(
        type=(float, type(None)),
        initial=None,
    )
    request_memory = field(
        type=(float, type(None)),
        initial=None,
    )
    request_disk = field(
        type=(float, type(None)),
        initial=None,
    )
```

File: task_processing/plugins/kubernetes/kubernetes_pod_executor.py (hunk 1 above)
Comment (anjaberens) [reply task_processing/plugins/kubernetes/kubernetes_pod_executor.py:454]: Its only for completeness. I can remove if you have strong feelings about this.

Comment (anjaberens) [reply task_processing/plugins/kubernetes/kubernetes_pod_executor.py:454]: I did try to set the default to None, iirc it caused the type system to get mad and make me do Union(None | float) kind of thing, hence this evasion.

File: task_processing/plugins/kubernetes/task_config.py (hunk 2 above)
Comment (anjaberens) [reply task_processing/plugins/kubernetes/task_config.py:316]: Flipping the name totally makes sense, i mostly worded it as request_x because of how it is in the structure itself.

File: task_processing/plugins/kubernetes/kubernetes_pod_executor.py (hunk 1 above)
Comment (nemacysts) [reply task_processing/plugins/kubernetes/kubernetes_pod_executor.py:454]: yea, I think without the `type=REAL_TYPE, type(None)` bit pyrsistent gets quite upset with None defaults :p

Comment (nemacysts) [reply task_processing/plugins/kubernetes/kubernetes_pod_executor.py:454]: yea, I say we can circle back and add this later if we ever end up needing this - it makes me a tad nervous as-is tho (mostly 'cause I'm not sure if the kubelet will protect us by evicting things when the node is under disk pressure or if we'll just run the node out of disk and kill it :p)

Comment (anjaberens) [reply task_processing/plugins/kubernetes/kubernetes_pod_executor.py:454]: I set this up and it works, main wrinkle is without the factory it gets upset when it gets a integer value (ie 768) for the fields. I'm not sure how much we care about this or should i go ahead and create a float_or_none factory?

Comment (nemacysts) [reply task_processing/plugins/kubernetes/kubernetes_pod_executor.py]: that sounds pretty reasonable! i can definitely see folks passing in integer values for these

Comment (anjaberens) [reply task_processing/plugins/kubernetes/kubernetes_pod_executor.py]: For the records - the testing this on jolt works. I will follow up next week with a factory for this.

**Review by nemacysts — APPROVED** [review]

File: task_processing/plugins/kubernetes/task_config.py [hunk 3]
```diff
@@ -238,6 +238,12 @@ def _valid_service_account_name(
     )
 
 
+def _float_or_none(str: Optional[str]) -> Optional[float]:
+    if str:
+        return float(str)
+    return None
```
Comment (nemacysts) [inline task_processing/plugins/kubernetes/task_config.py:244]: should this be something like:
```suggestion
def _float_or_none(val: Any) -> Optional[float]:
    return float(val) if val is not None else None
```

otherwise, the type hints aren't exactly correct (`_float_or_none(1)` is a valid call, but `1` is not a string.

thinking about this line also made me realize that I think we want an `invariant` set for the `*_request` fields as they should always be non-zero if set (I was going to say that the current factory would return `None` if passed in `0` or `0.0`, but then I realized that we don't really want 0-valued requests in the first place)

that said, I'm not quite sure what the ordering is here (i.e., is `factory()` called first or is `invariant()`?) so we'll probably need to do some testing here

Comment (anjaberens) [reply task_processing/plugins/kubernetes/task_config.py]: seems to run the invariant after factory i believe (?)

**Review by nemacysts — APPROVED** [review]
